runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,300 @@
+
+import java.util.*;
+
+/**
+ * Persisted Queue supported by CorfuDB using distributed State Machine Replication.
+ * Entries enqueued are backed by a LinkedHashMap where each entry is mapped to a unique and
+ * globally ordered generated id that is returned upon successful <b>enqueue()</b>.
+ * <b>entryList()</b> returns the enqueued entries along with their ids sorted by
+ * the order in which their <b>enqueue()</b> operations materialized.
+ * Instead of a dequeue() this Queue supports a <b>remove()</b> which accepts the id of the element.
+ * Entries cannot be modified in-place and but can be removed from anywhere in the queue.
+ *
+ * Created by hisundar on 5/8/19.
+ *
+ * @param <E>   Type of the entry to be enqueued into the persisted queue
+ */
+@Slf4j
+@CorfuObject
+public class CorfuQueue<E> implements ISMRObject {
+    /**
+     * The "main" linked map which contains the primary key-value mappings.
+     */
+    private final Map<Long, E> mainMap = new LinkedHashMap<>();
+
+    private final OrderedGuidGenerator guidGenerator;
+
+    public CorfuQueue(OrderedGuidGenerator guidGenerator) {
+        this.guidGenerator = guidGenerator;
+    }
+
+    /**
+     * Returns the size of the queue at a given point in time.
+     */
+    @Accessor
+    public int size() {
+        return mainMap.size();
+    }
+
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the order of insertions cannot be guaranteed
+     * unless a transaction is used.
+     * Capacity restrictions and backoffs must be implemented outside this
+     * interface. Consider validating the size of the queue against a high
+     * watermark before enqueue.
+     *
+     * @param e the element to add
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this queue
+     * @return Globally ordered and unique ID representing this entry in the persistent queue
+     * WARNING: The ID returned by this function will NOT be the comparable with the ID
+     *          returned by the CorfuQueueRecord from entryList() method because if this
+     *          method executes within a transaction, the comparable ordering is not
+     *          known until the transaction commits.
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    @TransactionalMethod
+    public UUID enqueue(E e) {
+        final Long id = guidGenerator.nextLong();
+        this.put(id, e);
+        return new UUID(0, id);
+    }
+
+    /**
+     * CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> {
```

Comment:
If you return an ordered map, you can get rid of this class. I don't see a strong reason to transform the map to a List when the map is already ordered.

Reply:
There are a few reasons:
1. The Map internally is a long=>entry, to return something consumable externally it will have to be remapped to an UUID=>entry, so the internal map to external consumption still needs an iterator.
2. I was addressing comments from @xnull who asked me to add this class to avoid overloading the Map.